src/core/mesh/qgsmeshlayer.cpp

```diff
@@ -352,6 +352,122 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
   return block3d;
 }
 
+QgsMeshDatasetValue QgsMeshLayer::dataset1DValue( const QgsMeshDatasetIndex &index, const QgsRectangle &rectangle ) const
+{
+  QgsMeshDatasetValue value;
+  QgsPointXY centerPoint = rectangle.center();
+  const QgsTriangularMesh *mesh = triangularMesh();
+  if ( mesh &&
+       mesh->contains( QgsMesh::Edge ) &&
+       mDataProvider->isValid() &&
+       index.isValid() )
+  {
+    // search for the closest edge in rectangle from point
+    const QList<int> edgeIndexes = mesh->edgeIndexesForRectangle( rectangle );
+    int selectedIndex = -1;
+    double sqrMaxDistFromPoint = pow( rectangle.width() / 2, 2 );
+    QgsPointXY projectedPoint;
+    for ( const int edgeIndex : edgeIndexes )
+    {
+      const QgsMeshEdge &edge = mesh->edges().at( edgeIndex );
+      const QgsMeshVertex &vertex1 = mesh->vertices()[edge.first];
+      const QgsMeshVertex &vertex2 = mesh->vertices()[edge.second];
+      QgsPointXY projPoint;
+      double sqrDist = centerPoint.sqrDistToSegment( vertex1.x(), vertex1.y(), vertex2.x(), vertex2.y(), projPoint );
+      if ( sqrDist < sqrMaxDistFromPoint )
+      {
+        selectedIndex = edgeIndex;
+        projectedPoint = projPoint;
+      }
+    }
+
+    if ( selectedIndex >= 0 )
+    {
+      const QgsMeshDatasetGroupMetadata::DataType dataType = dataProvider()->datasetGroupMetadata( index ).dataType();
+      switch ( dataType )
+      {
+        case QgsMeshDatasetGroupMetadata::DataOnEdges:
+        {
+          value = dataProvider()->datasetValue( index, selectedIndex );
+        }
+        break;
+
+        case QgsMeshDatasetGroupMetadata::DataOnVertices:
+        {
+          const QgsMeshEdge &edge = mesh->edges()[selectedIndex];
+          const int v1 = edge.first, v2 = edge.second;
+          const QgsPoint p1 = mesh->vertices()[v1], p2 = mesh->vertices()[v2];
+          const QgsMeshDatasetValue val1 = dataProvider()->datasetValue( index, v1 );
+          const QgsMeshDatasetValue val2 = dataProvider()->datasetValue( index, v2 );
+          double length = p1.distance( p2 );
+          const double x = val1.x() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.x() - val1.x() ) / length;
+          double y = std::numeric_limits<double>::quiet_NaN();
+          bool isVector = dataProvider()->datasetGroupMetadata( index ).isVector();
+          if ( isVector )
+            y = val1.y() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.y() - val1.y() ) / length;;
```

Comment:
p1.distance( projectedPoint.x(), projectedPoint.y() ) can be used from calculation of x?